keylime/crypto.py

```diff
@@ -79,65 +79,81 @@ def strbitxor(a,b):
     for i in range(len(a)):
         retval[i] = a[i] ^ b[i]
     return retval
- 
+
 def kdf(password,salt):
     return KDF.PBKDF2(password, salt, dkLen=32, count=2000)
-     
+
 def do_hmac(key,value):
-    h = HMAC.new(key,str(value),digestmod=SHA384.new())
+    value = value.encode('utf-8')
+    # Let's only encode if its not a byte
+    try:
+        key = key.encode('utf-8')
+    except AttributeError:
+        pass
+
+    h = HMAC.new(key,value,digestmod=SHA384.new())
     return h.hexdigest()
- 
+
+def sha2(value):
+    return SHA384.new(data=value).hexdigest()
+
 def _pad(s):
     '''
     Returns the string padded with its length such
     that is a multiple of 16
     Appends 10* at the bit level. Following ISO/IEC 9797-1
     - padding mode 2
     '''
+    # Let's only encode if its not a byte
+    try:
+        s = s.encode('utf-8') #
+    except AttributeError:
+        pass
     pad_len = AES.block_size - (len(s) % AES.block_size) - 1
-    padding = chr(0x80)+'\0'*pad_len
+    padding = b'\x80'+b'\0'*pad_len
+    result = s + padding
     return s + padding
- 
+
 def _strip_pad(s):
     '''
     Strips the padding from the string
     '''
     return s.rstrip(b'\0')[:-1]
```

Comment:
This isn't relevant to the py3 conversion, but shouldn't we be stripping the appended \x80 character as well @nabilschear ?  Under padding mode 2, we add a 1 (always) and then pad with 0's until we hit a multiple of 16 (if necessary).  

In python2 this might not make a difference, since `chr(0x80)` isn't valid 7-bit ASCII and isn't printable, but this might cause some unexpected problems now that we're using UTF-8.

Reply:
Do you know if this needs to be updated, @nabilschear?

Reply:
the `[:-1]` strips out the \x80 char after removing all the zeros.